Calc degrees

gotdairyya opened the pull request:
### Does this PR close any open issues?
Closes #386 

### Give a longer description of what this PR addresses and why it's needed
Let's the user filter nodes based on the number of out degrees.

### Are there any additional TODOs before this PR is ready to go?
I'm sure the code will need some sprucing up. I'm not sure I computed the filtering in the best way.
TODOs:
- [x] Review code for filtering
- [x] Check how this works with aggregated networks and queried networks

gotdairyya commented:
Also, @JackWilb I set the min and max attributes for the range-slider, but for some reason, they aren't limits (ex. if the min in 0, I can still input a value less than 0)

JackWilb reviewed (changes requested):
Looking good. I pushed a change for the slider UI (which I think will address your most recent comment) and I left some comments for how you might improve the filtering. I also noticed some general stuff that it would be best to fix up. Let me know if you'd like me to help with any of it

gotdairyya commented on src/store/index.ts:
```diff
@@ -456,13 +461,14 @@ const {
       };
       commit.setAttributeValues(networkElements);
+      commit.setNetworkOnLoad(networkElements);
+      commit.setDegreeRange([0, networkElements.nodes.length]);
```
I would normally agree but in my example network there is a node with no interactions for this specific dataset so by that metric (max degree) it would never show up

gotdairyya commented on src/lib/utils.ts:
```diff
@@ -26,6 +26,9 @@ export function defineNeighbors(nodes: any[], edges: Edge[]) {
   });
 
+  // eslint-disable-next-line no-return-assign, no-param-reassign
+  nodes.forEach((d) => d.degreeCount = new Set(d.neighbors).size);
```
> Is this how the degree count should work? If there are 100 edges between 2 people, they would have degree 1

Did I misinterpret what `d.neighbors` represents?


> Another issue is that we're tracking the same data in 2 different variables. It seems that degree count is just the size of the neighbors. Why don't we make neighbors a set or an object (to count edges, e.g. {"Jack": 10, "Derya": 20}), and then just use the size or length of keys instead of making a separate degreeCount variable?

Why would that be better? If it is a node variable then it will also show up in the UpSet table which could be useful.

JackWilb replied:
Derya showed me that this is a bad idea. We've pivoted a bit from this

gotdairyya commented on src/components/Controls.vue:
```diff
@@ -78,6 +79,62 @@ export default defineComponent({
     const maxConnections = computed(() => store.state.maxConnections);
 
+    // Calculate node degrees
+    const degreeMax = computed(() => (networkOnLoad.value !== null ? networkOnLoad.value.nodes.length : 0));
```
I will check

gotdairyya commented on src/components/Controls.vue:
```diff
@@ -78,6 +79,62 @@ export default defineComponent({
+          // Remove edges that don't match degree criteria
+          const edge = newNetwork.edges[i];
+          // eslint-disable-next-line radix
+          if (nodeDegreeDict[edge._from] < parseInt(degreeRange.value[0]) || nodeDegreeDict[edge._from] > parseInt(degreeRange.value[1])) {
```
`degreeRange` is typecast for `number[]` but when I was testing the code and console logging, sometimes it would return a string. I wasn't sure why it was doing that

gotdairyya commented on src/components/Controls.vue:
```diff
@@ -78,6 +79,62 @@ export default defineComponent({
+      if (networkOnLoad.value !== null && degreeRange.value[0] === 0 && degreeRange.value[1] === networkOnLoad.value.nodes.length) {
+        store.dispatch.updateNetwork({ network: networkOnLoad.value });
+      } else {
+        const newNetwork = JSON.parse(JSON.stringify(networkOnLoad.value));
```
:thumbs_up:

gotdairyya replied:
Had to use `// eslint-disable-next-line no-undef`, tracking in issue #390

gotdairyya commented on src/components/Controls.vue:
```diff
@@ -78,6 +79,62 @@ export default defineComponent({
+        newNetwork.nodes.forEach((node: Node) => nodeDegreeDict[node._id] = node.degreeCount);
+
+        // eslint-disable-next-line no-plusplus
+        for (let i = 0; i < newNetwork.edges.length; i++) {
```
oh that's much nicer, thank you!

gotdairyya replied:
done

gotdairyya commented on src/components/Controls.vue:
```diff
@@ -78,6 +79,62 @@ export default defineComponent({
+        }
+
+        // eslint-disable-next-line no-plusplus
+        for (let i = 0; i < newNetwork.nodes.length; i++) {
```
done

gotdairyya commented:
@JackWilb I've updated how the degrees are calculated. Could you help with the type setting? And it should be good after that unless you notice anything else

JackWilb commented on src/components/Controls.vue:
```diff
@@ -78,6 +81,53 @@ export default defineComponent({
+
+    const degreeMax = computed(() => {
+      let maxCounter = 0;
+      if (networkOnLoad.value != null) {
```
This is null on load for me, which causes the slider to initialize to [0,0]

gotdairyya replied:
I think I fixed this

JackWilb replied:
It seems so!

JackWilb commented on src/store/index.ts:
```diff
@@ -361,6 +363,43 @@ const {
+    setNodeDegreeDict(state) {
+      // Create dict of row nodes and max degrees
+      if (state.networkOnLoad != null) {
+        state.networkOnLoad.nodes.forEach((node) => {
```
Should this be `networkOnLoad`? If we make it `network`, then it might work better with aggregation, etc. We'd just make sure we call it on load and when aggregating.

gotdairyya replied:
we want it to be `networkOnLoad` because the filtering updates the `network` and that's how the matrix is redrawn. I can check for aggregation and change the network that is used (same for querying). 

If you have another idea for how to redraw the matrix without updating the network that would be great.

JackWilb commented on src/store/index.ts:
```diff
@@ -361,6 +363,43 @@ const {
+
+        // Remove edges that don't match degree criteria
+        newNetwork.edges = newNetwork.edges.filter((edge: Edge) => {
+          if (state.nodeDegreeDict[edge._from] >= degreeRange[0] && state.nodeDegreeDict[edge._from] <= degreeRange[1]) {
```
This gives us all out edges from nodes that are in the degree range (and all their neighbors). Should we also allow all in edges (especially in the case that we don't care about edge directionality)?

JackWilb replied:
Also, should we include their neighbors or only include them if they are also in the degree range?

gotdairyya replied:
> This gives us all out edges from nodes that are in the degree range (and all their neighbors). Should we also allow all in edges (especially in the case that we don't care about edge directionality)?

Oh, yeah I can do that based on whether it is a directional matrix or not

- [ ] Account for directionality in filtering

gotdairyya replied:
> Also, should we include their neighbors or only include them if they are also in the degree range?

I don't understand this, sorry

JackWilb replied:
> > Also, should we include their neighbors or only include them if they are also in the degree range?
> 
> I don't understand this, sorry

You add both nodes to `nodeSet` when the from node has the right degree. That means the to node will get included. Is that what we want? Or should we just show the connections between all nodes that have the right degree?

That would change the logic to:
1. Filter the nodes to include only those in the degree range.
2. Find all edges that have from and to in the filtered nodes.

I guess I'm just wondering if we're doing this filtering step backwards as it's written

gotdairyya replied:
So this is what I did at first and then I realized that we need to track the to nodes or else we may end up having an edge without an edge that is connected to and that will throw an error when the matrix is drawn.

JackWilb replied:
The edge should only be included if both the from and to are in the selected subset. That should fix that problem

JackWilb commented on src/components/FilterOverlay.vue:
```diff
@@ -159,8 +162,17 @@ export default defineComponent({
           if (aqlNetwork.nodes.length !== 0) {
-          // Update state with new network
-            store.dispatch.updateNetwork({ network: aqlNetwork });
+            const nodes = defineNeighbors(aqlNetwork.nodes, aqlNetwork.edges as Edge[]);
```
It feels weird to have the `defineNeighbors` step here. I think this should just be in the store in the update network function. Let me know what you think about that

gotdairyya replied:
I don't think so! It is another source of creating new network data which I see as equivalent to when smaller networks are loaded

gotdairyya replied:
> This code is looking pretty good so far. I have a couple of suggestions in the comments, but the ones about the computed vs a ref with a watcher might not work.
> 
> On another note, the node degree filtering doesn't work with an aggregated network. It also doesn't work with quering, since it just copies networkOnLoad, which is not updated. I'm not sure the best way to make all of this work with that 😅 but maybe it will become obvious as you play some more with it... If not, I'm happy to dig in and see what I can see for simplifying it all

To Do:

- [ ] have filtering work on aggregated network
- [ ] have filtering work on queried network

JackWilb replied:
I wasn't very clear here. What I meant was, don't we want to redefine the neighbors each time we update a network? If so, it would make sense to do that in the store function that handles it. However, it could be that we don't want to do it every time, and that there's a good reason to do it here (I'm thinking aggregation)

JackWilb commented on src/components/FilterOverlay.vue:
```diff
@@ -159,8 +162,17 @@ export default defineComponent({
+            };
+
+            // Update state with new network
+            store.commit.setNetworkOnLoad(aqlNetworkElements);
+            store.dispatch.updateNetwork({ network: aqlNetworkElements });
```
If you accept the other change in this file, these will need to be switched and you'll need to pass another network to `setNetworkOnLoad`
```suggestion
            store.dispatch.updateNetwork({ network: aqlNetworkElements });
            store.commit.setNetworkOnLoad(store.state.network);
```

JackWilb commented on src/components/Controls.vue:
```diff
@@ -77,6 +77,12 @@ export default defineComponent({
     });
     const maxConnections = computed(() => store.state.maxConnections);
+    const maxDegree = computed(() => store.state.maxDegree);
+    const degreeRange = ref([0, maxDegree.value]);
+
+    watch([maxDegree], () => {
+      degreeRange.value = [0, maxDegree.value];
+    });
```
Can this be made into a computed? I'm not sure if this will have issues with reactivity. It could also cause issues with the value in the html below.

```suggestion
    const degreeRange = computed(() => [0, maxDegree.value]);
```

JackWilb commented on src/components/Controls.vue:
```diff
@@ -375,6 +388,36 @@ export default defineComponent({
+          <v-list-item>
+            <v-list-item-content> Degree </v-list-item-content>
+            <v-range-slider
+              v-model="degreeRange"
```
I wonder if this `v-model` is needed if you switch to using a computed.

JackWilb reviewed (changes requested):
This code is looking pretty good so far. I have a couple of suggestions in the comments, but the ones about the computed vs a ref with a watcher might not work.

On another note, the node degree filtering doesn't work with an aggregated network. It also doesn't work with quering, since it just copies networkOnLoad, which is not updated. I'm not sure the best way to make all of this work with that 😅 but maybe it will become obvious as you play some more with it... If not, I'm happy to dig in and see what I can see for simplifying it all

gotdairyya commented:
@JackWilb I made quite a few changes so I will list them out here. I didn't make the ref to a computed, I had tried that and had some issues, if you can look at that that would be great.

**Major Changes:**
- Moved the computation of the node degree dict to the utils
- Added logic for directionality in the matrix
- Filtering works for any combination of aggregated network and/or queried network

gotdairyya commented:
Jack and I decided to go ahead and disable degree filtering when aggregating the network because both of these methods are meant to make the network smaller. It also removes visual complexity.

**Left to do:**
- [ ] Fix color for filtered children cells
- [ ] Reset rangeSlider when aggregating
- [ ] Remove filtered nodes from aggregation function
- [ ] Check if new design works with queried network (or @JackWilb do we disable it here too)

JackWilb commented:
> Remove filtered nodes from aggregation function

We could probably reset the network before aggregation

> Check if new design works with queried network (or @JackWilb do we disable it here too)

Querying should reset everything, and give you the queried network

JackWilb reviewed (dismissed):
Looks good! I have a couple of suggestions for you to fix up before you merge. It should simplify some checks and remove some duplicated code

gotdairyya commented on src/components/Controls.vue:
```diff
@@ -185,6 +193,12 @@ export default defineComponent({
     function aggregateNetwork(varName: string) {
+      if (filtered.value) {
+        store.commit.setFilteredNetwork(false);
+        if (networkOnLoad.value !== null) { store.dispatch.updateNetwork({ network: networkOnLoad.value }); }
```
Done

gotdairyya commented on src/lib/utils.ts:
```diff
@@ -55,3 +55,31 @@ export function formatShortDate(date: Date) {
+  if (networkPreFilter != null || networkOnLoad !== null) {
+    baseNetwork = queried ? structuredClone(networkPreFilter) : structuredClone(networkOnLoad);
+  }
+  if (baseNetwork !== null) {
```
done

gotdairyya commented on src/store/index.ts:
```diff
@@ -362,6 +376,88 @@ const {
+
+    setDegreeNetwork(state, degreeRange: number[]) {
+      // Determine correct network to use
+      let baseNetwork: Network | null = { nodes: [], edges: [] };
```
done

gotdairyya commented on src/lib/utils.ts:
```diff
@@ -55,3 +55,31 @@ export function formatShortDate(date: Date) {
+    baseNetwork.edges.forEach((edge: Edge) => {
+      if (directionalEdges) {
+        // eslint-disable-next-line no-unused-expressions
+        Object.prototype.hasOwnProperty.call(nodeDegreeDict, edge._from) ? nodeDegreeDict[edge._from] += 1 : nodeDegreeDict[edge._from] = 1;
```
Opened in issue #406
